crates/accelerate/src/unitary_synthesis.rs

```diff
@@ -0,0 +1,1082 @@
+    TwoQubitBasisDecomposer, TwoQubitGateSequence, TwoQubitWeylDecomposition,
+};
+use crate::QiskitError;
+
+const PI2: f64 = PI / 2.;
+const PI4: f64 = PI / 4.;
+
+#[derive(Clone, Debug)]
+enum DecomposerType {
+    TwoQubitBasisDecomposer(Box<TwoQubitBasisDecomposer>),
+    XXDecomposer(PyObject),
+}
+
+struct DecomposerElement {
+    decomposer: DecomposerType,
+    gate: NormalOperation,
+}
+
+#[derive(Clone, Debug)]
+struct TwoQubitUnitarySequence {
+    gate_sequence: TwoQubitGateSequence,
+    decomp_gate: NormalOperation,
+}
+
+// Used in get_2q_decomposers. If the found 2q basis is a subset of GOODBYE_SET,
+// then we know TwoQubitBasisDecomposer is an ideal decomposition and there is
+// no need to bother trying the XXDecomposer.
+static GOODBYE_SET: [&str; 3] = ["cx", "cz", "ecr"];
+
+fn get_target_basis_set(target: &Target, qubit: PhysicalQubit) -> EulerBasisSet {
+    let mut target_basis_set: EulerBasisSet = EulerBasisSet::new();
+    let target_basis_list = target.operation_names_for_qargs(Some(&smallvec![qubit]));
+    match target_basis_list {
+        Ok(basis_list) => {
+            EULER_BASES
+                .iter()
+                .enumerate()
+                .filter_map(|(idx, gates)| {
+                    if !gates.iter().all(|gate| basis_list.contains(gate)) {
+                        return None;
+                    }
+                    let basis = EULER_BASIS_NAMES[idx];
+                    Some(basis)
+                })
+                .for_each(|basis| target_basis_set.add_basis(basis));
+        }
+        Err(_) => target_basis_set.support_all(),
+    }
+    if target_basis_set.basis_supported(EulerBasis::U3)
+        && target_basis_set.basis_supported(EulerBasis::U321)
+    {
+        target_basis_set.remove(EulerBasis::U3);
+    }
+    if target_basis_set.basis_supported(EulerBasis::ZSX)
+        && target_basis_set.basis_supported(EulerBasis::ZSXX)
+    {
+        target_basis_set.remove(EulerBasis::ZSX);
+    }
+    target_basis_set
+}
+
+fn apply_synth_dag(
+    py: Python<'_>,
+    out_dag: &mut DAGCircuit,
+    out_qargs: &[Qubit],
+    synth_dag: &DAGCircuit,
+) -> PyResult<()> {
+    for out_node in synth_dag.topological_op_nodes()? {
+        let mut out_packed_instr = synth_dag.dag()[out_node].unwrap_operation().clone();
+        let synth_qargs = synth_dag.get_qargs(out_packed_instr.qubits);
+        let mapped_qargs: Vec<Qubit> = synth_qargs
+            .iter()
+            .map(|qarg| out_qargs[qarg.0 as usize])
+            .collect();
+        out_packed_instr.qubits = out_dag.qargs_interner.insert(&mapped_qargs);
+        out_dag.push_back(py, out_packed_instr)?;
+    }
+    out_dag.add_global_phase(py, &synth_dag.get_global_phase())?;
+    Ok(())
+}
+
+fn apply_synth_sequence(
+    py: Python<'_>,
+    out_dag: &mut DAGCircuit,
+    out_qargs: &[Qubit],
+    sequence: &TwoQubitUnitarySequence,
+) -> PyResult<()> {
+    let mut instructions = Vec::new();
```

Comment:
This is always the size of the gate sequence right?
```suggestion
    let mut instructions = Vec::with_capacity(sequence.gate_sequence.gates().len());
```